trufflesuite/truffle
# Internal improvement: get husky and lint-staged working

This PR updates husky to v6 and hooks up lint-staged which has stopped working for some reason (#3530). It uses the newest version of Husky which relies on better git api introduced in 2016 ([See this blog post](https://blog.typicode.com/husky-git-hooks-javascript-config/) for details), as well as building on the work of @bergarces and @haltman-at wrt eslint configuration. 

Also cherry picked @haltman-at 's work around to fix a webpack `rx-lite-aggregates` issue.
<details> <summary>See details</summary>

```
truffle: ERROR in ./node_modules/rx-lite-aggregates/rx.lite.aggregates.js
truffle: Module not found: Error: Can't resolve './rx.lite' in '/home/amal/work/truffle/node_modules/rx-lite-aggregates'
truffle: resolve './rx.lite' in '/home/amal/work/truffle/node_modules/rx-lite-aggregates'
truffle:   using description file: /home/amal/work/truffle/node_modules/rx-lite-aggregates/package.json (relative path: .)
truffle:     using description file: /home/amal/work/truffle/node_modules/rx-lite-aggregates/package.json (relative path: ./rx.lite)
truffle:       no extension
truffle:         /home/amal/work/truffle/node_modules/rx-lite-aggregates/rx.lite doesn't exist
truffle:       .js
truffle:         /home/amal/work/truffle/node_modules/rx-lite-aggregates/rx.lite.js doesn't exist
truffle:       .json
truffle:         /home/amal/work/truffle/node_modules/rx-lite-aggregates/rx.lite.json doesn't exist
truffle:       .wasm
truffle:         /home/amal/work/truffle/node_modules/rx-lite-aggregates/rx.lite.wasm doesn't exist
truffle:       as directory
truffle:         /home/amal/work/truffle/node_modules/rx-lite-aggregates/rx.lite doesn't exist
truffle:  @ ./node_modules/inquirer/lib/ui/prompt.js 3:9-38
truffle:  @ ./node_modules/inquirer/lib/inquirer.js 18:10-32
truffle:  @ ./packages/core/lib/services/analytics/google.js 15:17-36
truffle:  @ ./packages/core/lib/services/analytics/main.js 2:18-40
truffle: webpack 5.21.2 compiled with 1 error and 27 warnings in 23499 ms
truffle: error Command failed with exit code 1.
truffle: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
truffle: error Command failed with exit code 1.
truffle: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
truffle: error Command failed with exit code 1.
truffle: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run prepare exited 1 in 'truffle'
```
</details>

## TODO
- [ ] investigate how this will work on windows because the husky scripts are all bash :/ . This should not hold back merging. That's a problem for the windows crew

## Note to maintainers

These are necessary one-time steps to install and configure husky

1. Cleanup your `.git/hooks folder` which has git hook definitions for a prior version of husky. 
    ```console
    cd $TRUFFLE_ROOT/.git
    grep -l husky * | xargs rm
    ```
1. Run `yarn prepare` to install the husky hooks

## How to test
Commit changes to ts, or js files and see if the hooks kick in and does appropriate linting and prettier things. N.B. you can undo the last commit in git with the following command: 
```
git reset HEAD~1
```
